src/systemcmds/i2c_launcher/i2c_launcher.cpp

```diff
@@ -0,0 +1,221 @@
+/****************************************************************************
+ *
+ *   Copyright (C) 2023 PX4 Development Team. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ * 3. Neither the name PX4 nor the names of its contributors may be
+ *    used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+#include <stdlib.h>
+#include "i2c_launcher.hpp"
+#include <px4_platform_common/time.h>
+#include <px4_platform_common/log.h>
+#include <px4_arch/i2c_hw_description.h>
+
+constexpr I2CLauncher::I2CDevice I2CLauncher::_devices[];
+
+I2CLauncher::I2CLauncher() :
+	ModuleParams(nullptr),
+	ScheduledWorkItem(MODULE_NAME, px4::wq_configurations::lp_default)
```

Comment:
From a scheduling and locking perspective this would be a bit cleaner if running from the relevant I2Cx work queue.

Reply:
Hm, but that would be 4 queues, right?

Reply:
For whichever bus is relevant. Then from a scheduling perspective it should do something quick and minimal (eg check one address this cycle), then get out of the way for any real drivers already running.

Reply:
Note that it doesn't run when armed.

Reply:
We still need to be careful not to disrupt any other drivers running on that bus prearm.

Reply:
But if this runs for all external buses, it would schedule itself on multiple different work queues? Can you change the work queues on to easily like that?

Reply:
I would just keep it simple with one instance running per external bus, and only optimize that later if we need to.

Reply:
Done.